client/config/config.go

```diff
@@ -289,9 +355,406 @@ func (c *ClientTemplateConfig) Copy() *ClientTemplateConfig {
+		c.MaxStale == nil &&
+		c.MaxStaleHCL == "" &&
+		c.Wait.IsEmpty() &&
+		c.ConsulRetry.IsEmpty() &&
+		c.VaultRetry.IsEmpty()
+}
+
+// WaitConfig is mirrored from templateconfig.WaitConfig because we need to handle
+// the HCL conversion which happens in agent.ParseConfigFile
+// NOTE: Since Consul Template requires pointers, this type uses pointers to fields
+// which is inconsistent with how Nomad typically works. This decision was made
+// to maintain parity with the external subsystem, not to establish a new standard.
+type WaitConfig struct {
+	Enabled *bool          `hcl:"enabled,optional"`
+	Min     *time.Duration `hcl:"-"`
+	MinHCL  string         `hcl:"min,optional" json:"-"`
+	Max     *time.Duration `hcl:"-"`
+	MaxHCL  string         `hcl:"max,optional" json:"-"`
+}
+
+// Copy returns a deep copy of the receiver.
+func (wc *WaitConfig) Copy() *WaitConfig {
+	if wc == nil {
+		return nil
+	}
+
+	nwc := new(WaitConfig)
+
+	if wc.Enabled != nil {
+		nwc.Enabled = &*wc.Enabled
+	}
+	if wc.Min != nil {
+		nwc.Min = &*wc.Min
+	}
+	if wc.Max != nil {
+		nwc.Max = &*wc.Max
+	}
+
+	return wc
+}
+
+// Equals returns the result of reflect.DeepEqual
+func (wc *WaitConfig) Equals(other *WaitConfig) bool {
+	return reflect.DeepEqual(wc, other)
+}
+
+// IsEmpty returns true if the receiver only contains an instance with no fields set.
+func (wc *WaitConfig) IsEmpty() bool {
+	if wc == nil {
+		return true
+	}
+	return wc.Equals(&WaitConfig{})
+}
+
+// Validate returns an error  if the receiver is nil or empty or if Min is greater
+// than Max the user specified Max, or if the user didn't specify a Max.
+func (wc *WaitConfig) Validate() error {
+	// If the config is nil or empty return false so that it is never assigned.
+	if wc == nil || wc.IsEmpty() {
+		return errors.New("wait config is nil or empty")
+	}
+
+	// If min is nil, return
+	if wc.Min == nil {
+		return nil
+	}
+
+	// If min isn't nil, make sure Max is less than Min.
+	if wc.Max != nil {
+		if *wc.Min > *wc.Max {
+			return fmt.Errorf("wait config min %d is greater than max %d", *wc.Min, *wc.Max)
+		}
+	}
+
+	// Otherwise, return nil. Consul Template will set a Max based off of Min.
+	return nil
+}
+
+// Merge merges two WaitConfigs. The passed instance always takes precedence.
+func (wc *WaitConfig) Merge(b *WaitConfig) *WaitConfig {
+	if wc == nil {
+		return b
+	}
+
+	result := *wc
+	if b == nil {
+		return &result
+	}
+
+	if b.Enabled != nil {
+		result.Enabled = &*b.Enabled
+	}
+
+	if b.Min != nil {
+		result.Min = &*b.Min
+	}
+
+	if b.MinHCL != "" {
+		result.MinHCL = b.MinHCL
+	}
+
+	if b.Max != nil {
+		result.Max = &*b.Max
+	}
+
+	if b.MaxHCL != "" {
+		result.MaxHCL = b.MaxHCL
+	}
+
+	return &result
+}
+
+// ToConsulTemplate converts a client WaitConfig instance to a consul-template WaitConfig
+// TODO: Needs code review. The caller (TaskTemplateManager) takes direct pointers
+// to other configuration values. Need to make sure that desired here as well.
```

Comment:
I just felt a little apprehension passing pointer references around rather than copies of pointer values since it breaks encapsulation. I will go ahead and remove this comment since 

1. It behaves like the rest of the fields
2. It gets loaded only on agent startup
3. TaskTemplateManager only reads and never writes